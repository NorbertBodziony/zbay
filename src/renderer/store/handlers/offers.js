import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import { DateTime } from 'luxon'
import { remote } from 'electron'
import { getVault } from '../../vault'
import identitySelectors from '../selectors/identity'
import offersSelectors from '../selectors/offers'
import directMessagesQueue from '../selectors/directMessagesQueue'
import { messages as zbayMessages } from '../../zbay'
import channelSelectors from '../selectors/channel'
import contactsSelectors from '../selectors/contacts'
import channelHandlers from './channel'
import contactsHandlers from './contacts'
import directMessagesQueueHandlers from './directMessagesQueue'
import { messageType, actionTypes } from '../../../shared/static'
import appSelectors from '../selectors/app'
import appHandlers from './app'

import * as Here from './offers'

export const Offer = Immutable.Record(
  {
    address: '',
    itemId: '',
    name: '',
    lastSeen: '',
    messages: Immutable.List(),
    newMessages: Immutable.List()
  },
  'Offer'
)
const initialState = Immutable.Map()

const setMessages = createAction(actionTypes.SET_OFFER_MESSAGES)
const addOffer = createAction(actionTypes.ADD_OFFER)
const cleanNewMessages = createAction(actionTypes.CLEAN_OFFER_NEW_MESSAGESS)
const setLastSeen = createAction(actionTypes.SET_OFFER_LAST_SEEN)
const appendMessages = createAction(actionTypes.APPEND_OFFER_MESSAGES)
const appendNewMessages = createAction(actionTypes.APPEND_NEW_OFFER_MESSAGES)
const setOfferMessageBlockTime = createAction(
  actionTypes.SET_OFFER_MESSAGE_BLOCKTIME
)

export const actions = {
  setMessages,
  addOffer,
  cleanNewMessages,
  setLastSeen,
  appendMessages,
  appendNewMessages,
  setOfferMessageBlockTime
}
const createOfferAdvert = ({ payload, history }) => async (
  dispatch,
  getState
) => {
  await dispatch(createOffer({ payload }))
  await dispatch(
    contactsHandlers.epics.updateDeletedChannelTimestamp({
      address: payload.id + payload.offerOwner,
      timestamp: 0
    })
  )
  history.push(
    `/main/offers/${payload.id + payload.offerOwner}/${payload.address}`
  )
}
const createOffer = ({ payload }) => async (dispatch, getState) => {
  const offers = offersSelectors.offers(getState())
  if (!offers.find(e => e.itemId === payload.id + payload.offerOwner)) {
    const identityId = identitySelectors.id(getState())
    await getVault().offers.importOffer(identityId, payload)
    await dispatch(Here.loadVaultContacts())
  }
}
export const loadVaultContacts = () => async (dispatch, getState) => {
  const identityId = identitySelectors.id(getState())
  const offersVault = await getVault().offers.listOffers(identityId)
  const offers = offersSelectors.offers(getState())
  offersVault.forEach(async offer => {
    if (!offers.find(e => e.itemId === offer.offerId)) {
      const newOffer = new Offer({
        address: offer.address,
        itemId: offer.offerId,
        name: offer.name,
        lastSeen: offer.lastSeen
      })
      await dispatch(addOffer({ newOffer }))
      const message = await getVault().adverts.getAdvert(
        offer.offerId.substring(0, 64)
      )
      await dispatch(appendMessages({ message, itemId: offer.offerId }))
    }
  })
}

export const initMessage = () => async (dispatch, getState) => {
  const identityId = identitySelectors.id(getState())
  const offers = offersSelectors.offers(getState())
  const identityAddress = identitySelectors.address(getState())
  await offers.forEach(async offer => {
    const vaultMessages = await getVault().offers.listMessages({
      identityId,
      offerId: offer.itemId
    })
    const contact = contactsSelectors.contact(offer.address)(getState())
    const vaultToDisplayableMessages = vaultMessages.map(msg => {
      return zbayMessages.vaultToDisplayableMessage({
        message: {
          ...msg.properties,
          message: JSON.parse(msg.properties['message']).text,
          shippingData: JSON.parse(msg.properties['message']).shippingData,
          tag: JSON.parse(msg.properties['message']).tag,
          offerOwner: JSON.parse(msg.properties['message']).offerOwner,
          sender: {
            replyTo: msg.properties.sender,
            username: msg.properties.senderUsername
          },
          createdAt: parseInt(msg.properties.createdAt),
          type: parseInt(msg.properties.type)
        },
        identityAddress,
        receiver: { replyTo: offer.address, username: contact.username }
      })
    })
    await dispatch(
      setMessages({
        messages: vaultToDisplayableMessages,
        itemId: offer.itemId
      })
    )
  })
}
const refreshMessages = id => async (dispatch, getState) => {
  const identityId = identitySelectors.id(getState())
  const offer = offersSelectors.offer(id)(getState())
  const identityAddress = identitySelectors.address(getState())
  const vaultMessages = await getVault().offers.listMessages({
    identityId,
    offerId: id
  })
  const contact = contactsSelectors.contact(offer.address)(getState())
  const vaultToDisplayableMessages = vaultMessages.map(msg => {
    return zbayMessages.vaultToDisplayableMessage({
      message: {
        ...msg.properties,
        message: JSON.parse(msg.properties['message']).text,
        shippingData: JSON.parse(msg.properties['message']).shippingData,
        tag: JSON.parse(msg.properties['message']).tag,
        offerOwner: JSON.parse(msg.properties['message']).offerOwner,
        sender: {
          replyTo: msg.properties.sender,
          username: msg.properties.senderUsername
        },
        createdAt: parseInt(msg.properties.createdAt),
        type: parseInt(msg.properties.type)
      },
      identityAddress,
      receiver: { replyTo: offer.address, username: contact.username }
    })
  })
  vaultToDisplayableMessages.forEach(msg => {
    if (!offer.messages.find(m => m.id === msg.id)) {
      dispatch(appendMessages({ message: msg, itemId: offer.itemId }))
    }
  })
}

const sendItemMessageOnEnter = event => async (dispatch, getState) => {
  const enterPressed = event.nativeEvent.keyCode === 13
  const shiftPressed = event.nativeEvent.shiftKey === true
  const privKey = identitySelectors.signerPrivKey(getState())
  const dmQueue = directMessagesQueue.queue(getState())
  const channel = channelSelectors.channel(getState()).toJS()
  const messageToSend = channelSelectors.message(getState())
  const currentMessage = dmQueue.find(
    dm =>
      dm.get('recipientAddress') === channel.address &&
      dm.message.get('type') === messageType.ITEM_BASIC
  )
  const currentMessageKey = dmQueue.findKey(
    dm =>
      dm.get('recipientAddress') === channel.address &&
      dm.message.get('type') === messageType.ITEM_BASIC
  )
  if (enterPressed && !shiftPressed) {
    event.preventDefault()
    const messageQueueLock = appSelectors.directMessageQueueLock(getState())
    let locked = false
    if (!messageQueueLock) {
      await dispatch(appHandlers.actions.lockDmQueue())
      locked = true
    }
    let message
    if (currentMessage !== undefined && locked) {
      await dispatch(
        directMessagesQueueHandlers.actions.removeMessage(currentMessageKey)
      )
      message = zbayMessages.createMessage({
        messageData: {
          type: zbayMessages.messageType.ITEM_BASIC,
          data: {
            itemId: channel.id.substring(0, 64),
            text:
              currentMessage.message.getIn(['message', 'text']) +
              '\n' +
              messageToSend
          }
        },
        privKey
      })
    } else {
      message = zbayMessages.createMessage({
        messageData: {
          type: zbayMessages.messageType.ITEM_BASIC,
          data: {
            itemId: channel.id.substring(0, 64),
            text: messageToSend
          }
        },
        privKey
      })
    }
    dispatch(
      directMessagesQueueHandlers.epics.addDirectMessage(
        {
          message,
          recipientAddress: channel.address,
          recipientUsername: channel.id.substring(64)
        },
        null,
        false
      )
    )
    dispatch(channelHandlers.actions.setMessage(''))
    if (locked) {
      dispatch(appHandlers.actions.unlockDmQueue())
    }
  }
}
const updateLastSeen = ({ itemId }) => async (dispatch, getState) => {
  const identity = identitySelectors.data(getState())
  const lastSeen = DateTime.utc()
  const unread = offersSelectors.newMessages(itemId)(getState()).size
  remote.app.badgeCount = remote.app.badgeCount - unread
  await getVault().offers.updateLastSeen({
    identityId: identity.get('id'),
    offerId: itemId,
    lastSeen
  })
  dispatch(setLastSeen({ itemId, lastSeen }))
  dispatch(cleanNewMessages({ itemId }))
}
export const epics = {
  createOfferAdvert,
  loadVaultContacts,
  initMessage,
  sendItemMessageOnEnter,
  // sendItemTransferMessage,
  refreshMessages,
  createOffer,
  updateLastSeen
}

export const reducer = handleActions(
  {
    [setMessages]: (state, { payload: { itemId, messages } }) =>
      state.update(itemId, Offer(), item =>
        item.set('messages', item.messages.concat(messages))
      ),
    [addOffer]: (state, { payload: { newOffer } }) =>
      state.merge({ [newOffer.itemId]: newOffer }),
    [cleanNewMessages]: (state, { payload: { itemId } }) =>
      state.update(itemId, Offer(), item =>
        item.set('newMessages', Immutable.List())
      ),
    [appendMessages]: (state, { payload: { itemId, message } }) =>
      state.update(itemId, Offer(), item =>
        item.set('messages', item.messages.push(message))
      ),
    [appendNewMessages]: (state, { payload: { itemId, message } }) =>
      state.update(itemId, Offer(), item =>
        item.set('newMessages', item.newMessages.push(message))
      ),
    [setLastSeen]: (state, { payload: { itemId, lastSeen } }) =>
      state.update(itemId, Offer(), item => item.set('lastSeen', lastSeen)),
    [setOfferMessageBlockTime]: (
      state,
      { payload: { itemId, messageId, blockTime } }
    ) =>
      state.update(itemId, Offer(), cm =>
        cm.update('messages', messages => {
          const index = messages.findIndex(msg => msg.id === messageId)
          return messages.setIn([index, 'blockTime'], blockTime)
        })
      )
  },
  initialState
)
export default {
  actions,
  epics,
  reducer,
  loadVaultContacts
}
