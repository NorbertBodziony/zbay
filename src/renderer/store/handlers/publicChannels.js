import Immutable from 'immutable'
import { createAction, handleActions } from 'redux-actions'
import * as R from 'ramda'
import appSelectors from '../selectors/app'
import appHandlers from './app'
import channelsSelectors from '../selectors/channels'
import feesSelectors from '../selectors/fees'
import nodeSelectors from '../selectors/node'
import { getClient } from '../../zcash'
import { errorNotification, successNotification } from './utils'
import identitySelectors from '../selectors/identity'
import notificationsHandlers from './notifications'
import { messages } from '../../zbay'
import { ADDRESS_TYPE } from '../../zbay/transit'
import txnTimestampsHandlers from '../handlers/txnTimestamps'
import txnTimestampsSelector from '../selectors/txnTimestamps'
import { getVault } from '../../vault'
import feesHandlers from '../handlers/fees'
import staticChannels from '../../zcash/channels'
import { messageType, actionTypes } from '../../../shared/static'

export const _PublicChannelData = Immutable.Record(
  {
    address: '',
    name: '',
    description: '',
    owner: '',
    timestamp: 0,
    keys: {}
  },
  'PublicChannelData'
)
export const initialState = Immutable.Map({})

export const setPublicChannels = createAction(actionTypes.SET_PUBLIC_CHANNELS)

export const actions = {
  setPublicChannels
}
export const fetchPublicChannels = () => async (dispatch, getState) => {
  try {
    const publicChannels = channelsSelectors.publicChannels(getState())
    let txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())
    const transfers = await getClient().payment.received(
      publicChannels.get('address')
    )
    if (
      transfers.length ===
      appSelectors.transfers(getState()).get(publicChannels.get('address'))
    ) {
      return
    } else {
      dispatch(
        appHandlers.actions.reduceNewTransfersCount(
          transfers.length - appSelectors.transfers(getState()).get(publicChannels.get('address'))
        )
      )
      dispatch(
        appHandlers.actions.setTransfers({
          id: publicChannels.get('address'),
          value: transfers.length
        })
      )
    }
    for (const key in transfers) {
      const transfer = transfers[key]
      if (!txnTimestamps.get(transfer.txid)) {
        const result = await getClient().confirmations.getResult(transfer.txid)
        await getVault().transactionsTimestamps.addTransaction(
          transfer.txid,
          result.time
        )
        await dispatch(
          txnTimestampsHandlers.actions.addTxnTimestamp({
            tnxs: { [transfer.txid]: result.time.toString() }
          })
        )
      }
    }
    txnTimestamps = txnTimestampsSelector.tnxTimestamps(getState())
    const sortedTransfers = transfers.sort(
      (a, b) => txnTimestamps.get(a.txid) - txnTimestamps.get(b.txid)
    )
    const registrationMessages = await Promise.all(
      sortedTransfers.map(transfer => {
        const message = messages.transferToMessage(transfer)
        return message
      })
    )
    const sortedMessages = registrationMessages
      .filter(msg => msg !== null)
      .sort((a, b) => txnTimestamps.get(a.id) - txnTimestamps.get(b.id))
    let minfee = 0
    let publicChannelsMap = Immutable.Map({})
    const network = nodeSelectors.network(getState())
    for (const msg of sortedMessages) {
      if (
        msg.type === messageType.CHANNEL_SETTINGS &&
        staticChannels.zbay[network].publicKey === msg.publicKey
      ) {
        minfee = parseFloat(msg.message.minFee)
      }
      if (!msg.spent.gte(minfee) || msg.type !== messageType.PUBLISH_CHANNEL) {
        continue
      }
      const updateChannelSettings = R.findLast(
        settingsMsg =>
          settingsMsg.type === messageType.CHANNEL_SETTINGS_UPDATE &&
          settingsMsg.publicKey === msg.publicKey &&
          settingsMsg.message.updateChannelAddress ===
            msg.message.channelAddress
      )(sortedMessages)
      const channel = _PublicChannelData({
        address: msg.message.channelAddress,
        name: msg.message.channelName,
        description: updateChannelSettings
          ? updateChannelSettings.message.updateChannelDescription
          : msg.message.channelDescription,
        owner: msg.publicKey,
        keys: { ivk: msg.message.channelIvk },
        timestamp: txnTimestamps.get(msg.id)
      })
      if (channel !== null && !publicChannelsMap.get(channel.name)) {
        publicChannelsMap = publicChannelsMap.merge({ [channel.name]: channel })
      }
    }
    await dispatch(feesHandlers.actions.setPublicChannelFee(minfee))
    await dispatch(setPublicChannels(publicChannelsMap))
  } catch (err) {
    console.warn(err)
  }
}
export const publishChannel = ({
  channelAddress,
  channelName,
  channelDescription,
  channelIvk
}) => async (dispatch, getState) => {
  const identityAddress = identitySelectors.address(getState())
  const privKey = identitySelectors.signerPrivKey(getState())
  const network = nodeSelectors.network(getState())
  const publicChannel = channelsSelectors.publicChannels(getState())
  const fee = feesSelectors.publicChannelfee(getState())
  const message = messages.createMessage({
    messageData: {
      type: messages.messageType.PUBLISH_CHANNEL,
      data: {
        channelName,
        channelAddress,
        channelIvk,
        channelDescription,
        networkType:
          network === 'testnet'
            ? ADDRESS_TYPE.SHIELDED_TESTNET
            : ADDRESS_TYPE.SHIELDED_MAINNET
      }
    },
    privKey: privKey
  })
  const transfer = await messages.messageToTransfer({
    message,
    address: publicChannel.get('address'),
    identityAddress,
    amount: fee
  })
  try {
    await getClient().payment.send(transfer)
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        successNotification({
          message: 'Your channel has been published.'
        })
      )
    )
  } catch (err) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: "Couldn't create channel, please check node connection."
        })
      )
    )
  }
}
export const epics = {
  fetchPublicChannels,
  publishChannel
}

export const reducer = handleActions(
  {
    [setPublicChannels]: (state, { payload: publicChannels }) => {
      return state.merge(publicChannels)
    }
  },
  initialState
)

export default {
  reducer,
  epics,
  actions
}
