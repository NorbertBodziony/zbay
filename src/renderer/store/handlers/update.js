import { ipcRenderer } from 'electron'
import { actionCreators } from './modals'
import { createMigrationFile } from './identity'
export const checkForUpdate = () => async (dispatch, getState) => {
  // create file on new update
  // Note it will recreate file on each new update so file will be up to date
  await dispatch(createMigrationFile())
  dispatch(actionCreators.openModal('applicationUpdate')())
}

export const startApplicationUpdate = () => async (dispatch, getState) => {
  ipcRenderer.send('proceed-update')
  dispatch(actionCreators.closeModal('applicationUpdate')())
}

export const declineUpdate = () => async (dispatch, getState) => {
  dispatch(actionCreators.closeModal('applicationUpdate')())
}

export const epics = {
  checkForUpdate,
  startApplicationUpdate,
  declineUpdate
}

export default {
  epics
}
