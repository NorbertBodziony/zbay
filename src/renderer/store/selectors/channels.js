import { createSelector } from 'reselect'

import zcashChannels from '../../zcash/channels'
import nodeSelectors from './node'

const ignoredChannels = network => [
  zcashChannels.registeredUsers[network].address,
  zcashChannels.channelOfChannels[network].address,
  zcashChannels.priceOracle[network].address
]
const store = s => s
const channels = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return state
      .get('channels')
      .updateIn(['data'], channel =>
        channel.filter(
          channel => !ignoredChannels(network).includes(channel.get('address'))
        )
      )
  }
)

const data = createSelector(channels, nodeSelectors.network, (ch, network) => {
  let channelsData = ch.data
  if (channelsData.size) {
    const storeId = channelsData.findIndex(
      ch => ch.get('address') === zcashChannels.store[network].address
    )
    const storeChannel = channelsData.get(storeId)
    channelsData = channelsData.delete(storeId).unshift(storeChannel)
    const zbayId = channelsData.findIndex(
      ch => ch.get('address') === zcashChannels.general[network].address
    )
    const zbayChannel = channelsData.get(zbayId)
    channelsData = channelsData.delete(zbayId).unshift(zbayChannel)
  }
  return channelsData
})
const loader = createSelector(channels, ch => ch.loader)
const errors = createSelector(channels, c => c.get('errors'))

const generalChannelId = createSelector(
  data,
  nodeSelectors.network,
  (ch, network) => {
    const generalChannel = ch.find(
      c => c.get('address') === zcashChannels.general[network].address
    )
    return generalChannel && generalChannel.get('id')
  }
)
const usersChannel = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return state
      .get('channels')
      .updateIn(['data'], channel =>
        channel.find(
          channel =>
            channel.get('address') ===
            zcashChannels.registeredUsers[network].address
        )
      ).data
  }
)
const priceOracleChannel = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return state
      .get('channels')
      .updateIn(['data'], channel =>
        channel.find(
          channel =>
            channel.get('address') ===
            zcashChannels.priceOracle[network].address
        )
      ).data
  }
)
const publicChannels = createSelector(
  store,
  nodeSelectors.network,
  (state, network) => {
    return state
      .get('channels')
      .updateIn(['data'], channel =>
        channel.find(
          channel =>
            channel.get('address') ===
            zcashChannels.channelOfChannels[network].address
        )
      ).data
  }
)

const channelById = id =>
  createSelector(data, ch => ch.find(c => c.get('id') === id))
const ownedChannels = createSelector(data, ch =>
  ch.filter(c => {
    return c.get('keys').get('sk')
  })
)

const lastSeen = id => createSelector(channelById(id), ch => ch.get('lastSeen'))

export default {
  channelById,
  generalChannelId,
  usersChannel,
  channels,
  loader,
  lastSeen,
  data,
  errors,
  ownedChannels,
  publicChannels,
  priceOracleChannel
}
