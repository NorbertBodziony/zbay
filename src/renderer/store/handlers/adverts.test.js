/* eslint import/first: 0 */
/* eslint new-cap: ["off", { "newIsCap": false }] */

jest.mock('../../zcash')

import Immutable from 'immutable'
import * as R from 'ramda'
import { DateTime } from 'luxon'

import create from '../create'
import advertHandlers from './adverts'
import { createChannel } from '../../testUtils'
import { mock as zcashMock } from '../../zcash'
import { IdentityState, Identity } from './identity'
import { ChannelState } from '../handlers/channel'
import { ChannelsState } from '../handlers/channels'

describe('Messages queue reducer handles', () => {
  let store = null
  const values = {
    tag: 'testtag',
    background: 2,
    title: 'testtitle',
    provideShipping: 1,
    amount: 123.1,
    description: 'random description',
    zec: 0.123,
    usd: 12
  }

  beforeEach(() => {
    jest.clearAllMocks()
    store = create({
      initialState: Immutable.Map({
        identity: IdentityState({
          data: Identity({
            id: 'test-id',
            address: 'test address',
            name: 'Saturn',
            balance: '33.583004',
            signerPrivKey:
              '879aff43df53606d8ae1219d9347360e7a30d1c2f141e14c9bc96bb29bf930cb'
          })
        }),
        channel: ChannelState({
          spentFilterValue: 38,
          id: 'test',
          address: 'testAddress',
          shareableUri: 'test://uri',
          members: 2,
          message: 'Message written in the input'
        }),
        channels: ChannelsState({
          data: Immutable.fromJS([createChannel('test')])
        })
      })
    })

    jest.spyOn(DateTime, 'utc').mockImplementation(() => {
      return DateTime.fromSeconds(123)
    })
  })
  describe('epics', () => {
    it('sends sends advert', async () => {
      const actionPromise = store.dispatch(
        advertHandlers.epics.handleSend({ values })
      )

      await actionPromise
      const result = zcashMock.requestManager.z_sendmany.mock.calls
      expect(
        R.sortBy(call => {
          const [, transfer] = call
          return transfer.address
        }, result)
      ).toMatchSnapshot()
    })
  })
})
