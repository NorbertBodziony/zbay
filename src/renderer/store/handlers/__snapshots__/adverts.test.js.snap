// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messages queue reducer handles epics sends sends advert 1`] = `
Array [
  Array [
    "test address",
    Array [
      Object {
        "address": "zs1testaddresstest",
        "amount": "0",
        "memo": "ff02a7f071688e85cf3130c31fb627c314f779a58ed1893430c23e5719a0ebe4373b035de9d461f4ea5189e5e9472d19594724d5a3b1804cca7738aaeabe47690093000000007b7465737474616700000002746573747469746c6500000000000000000000000000000000000000000000402800000000000072616e646f6d206465736372697074696f6e00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      },
    ],
    1,
    0.0000025,
  ],
]
`;
